qualtran/bloqs/basic_gates/su2_rotation.py

```diff
@@ -0,0 +1,158 @@
+        e^{i\lambda} \sin(\theta) & - \cos(\theta)
+        \end{pmatrix}
+    $$
+
+    where $s$ is the global phase shift.
+
+    Args:
+        theta: rotation angle $\theta$ in the above matrix.
+        phi: phase angle $\theta$ in the above matrix.
+        lambd: phase angle $\lambda$ in the above matrix.
+        global_shift: phase angle $\alpha$, i.e. apply a global phase shift of $e^{i \alpha}$.
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+        Motlagh and Wiebe. (2023). Equation 7.
+    """
+
+    theta: float
+    phi: float
+    lambd: float  # cannot use `lambda` as it is a python keyword
+    global_shift: float = 0
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build(q=1)
+
+    @cached_property
+    def rotation_matrix(self) -> NDArray[np.complex_]:
+        return np.exp(1j * self.global_shift) * np.array(
+            [
+                [
+                    np.exp(1j * (self.lambd + self.phi)) * np.cos(self.theta),
+                    np.exp(1j * self.phi) * np.sin(self.theta),
+                ],
+                [np.exp(1j * self.lambd) * np.sin(self.theta), -np.cos(self.theta)],
+            ]
+        )
+
+    @staticmethod
+    def from_matrix(mat: NDArray[np.complex_]) -> 'SU2RotationGate':
+        theta = np.arctan2(np.abs(mat[1, 0]), np.abs(mat[0, 0]))
+        if np.isclose(np.cos(theta), 0):
+            alpha = 0
+            phi = np.angle(mat[0, 1] / np.sin(theta))
+            lambd = np.angle(mat[1, 0] / np.sin(theta))
+        else:
+            alpha = np.angle(-mat[1, 1] / np.cos(theta))
+            if np.isclose(np.sin(theta), 0):
+                phi = np.angle(mat[0, 0] / np.cos(theta) * np.exp(-1j * alpha))
+                lambd = 0
+            else:
+                phi = np.angle(mat[0, 1] / np.sin(theta) * np.exp(-1j * alpha))
+                lambd = np.angle(mat[1, 0] / np.sin(theta) * np.exp(-1j * alpha))
+
+        return SU2RotationGate(theta, phi, lambd, alpha)
+
+    def add_my_tensors(
+        self,
+        tn: 'qtn.TensorNetwork',
+        tag: Any,
+        *,
+        incoming: Dict[str, 'SoquetT'],
+        outgoing: Dict[str, 'SoquetT'],
+    ):
+        import quimb.tensor as qtn
+
+        tn.add(
+            qtn.Tensor(
+                data=self.rotation_matrix,
+                inds=(outgoing['q'], incoming['q']),
+                tags=[self.short_name(), tag],
+            )
+        )
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', q: 'SoquetT') -> Dict[str, 'SoquetT']:
+        q = bb.add(ZPowGate(exponent=2, global_shift=0.5 + self.global_shift / (2 * np.pi)), q=q)
+        q = bb.add(ZPowGate(exponent=1 - self.lambd / np.pi, global_shift=-1), q=q)
+        q = bb.add(Ry(angle=2 * self.theta), q=q)
+        q = bb.add(ZPowGate(exponent=-self.phi / np.pi, global_shift=-1), q=q)
+        return {'q': q}
+
```

Comment:
Can we change the decomposition to use 3 rotations instead also return `TComplexity(rotations=3)` ? 


We currently use a cost of `ceil(1.149 * log2(1.0 / eps) + 9.2)` for synthesizing 1 rotation (used in `TComplexity`). This formula is derived from https://arxiv.org/pdf/1404.5320.pdf 

In the paper above, they show that any 2qubit rotation can written using 3 Z rotations 

<img width="500" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/3ed8fd65-8894-4fe3-b6d9-6c332c3f5849">


and then they give an algorithm to synthesize Z rotations using clifford + T gates 

<img width="500" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/e32175f8-17af-4a56-9778-87d749582f68">


with the average cost scaling as `ceil(1.149 * log2(1.0 / eps) + 9.2)`

<img width="500" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/960a700e-9862-4e5e-bd0a-2695f45ba8b6">

Reply:
You can return a `CirqGateAsBloq(cirq.GlobalPhaseGate())` instead to make it explicit. We should eventually add a native global phase Bloq to Qualtran but this should work in the meantime

Reply:
Added in https://github.com/quantumlib/Qualtran/pull/828